faunadb-java-dsl/src/main/java/com/faunadb/client/types/Decoder.java

```diff
@@ -0,0 +1,216 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.errors.FaunaException;
+import com.faunadb.client.types.Types.CollectionType;
+import com.faunadb.client.types.Types.MapType;
+import com.faunadb.client.types.Types.SimpleType;
+import com.faunadb.client.types.Value.BytesV;
+import com.faunadb.client.types.Value.NullV;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.lang.reflect.Array;
+import java.lang.reflect.Type;
+import java.util.*;
+
+import static com.google.common.base.Defaults.defaultValue;
+import static java.lang.String.format;
+import static java.lang.reflect.Modifier.isAbstract;
+
+/**
+ * FaunaDB {@link Value} to object decoder.
+ */
+public final class Decoder {
+  private Decoder() {
+  }
+
+  /**
+   * Decode a FaunaDB {@link Value} to a specified type.
+   * <p>
+   * <p>This method is useful if you need to decode
+   * collections types.</p>
+   * <p>
+   * <pre>{@code
+   *     ArrayV array = new ArrayV(Arrays.asList(new StringV("str1"), new StringV("str2")));
+   *
+   *     Result<List<String>> listStrings = Decoder.decode(array, Types.arrayListOf(String.class));
+   *     Result<Set<String>> setStrings = Decoder.decode(array, Types.hashSetOf(String.class));
+   * }</pre>
+   * <p>
+   * See {@link Types} for more information.
+   *
+   * @param <T>     The return type of the method.
+   * @param value   The FaunaDB {@link Value} to be decoded.
+   * @param dstType The {@link Type} in which value should be decoded.
+   * @return A {@link Result} instance of type {@link T}
+   */
+  public static <T> Result<T> decode(Value value, Type dstType) {
+    if (value == null || value == NullV.NULL)
+      return Result.fail("Value is null");
+
+    return (Result<T>) decodeImpl(value, Types.of(dstType));
```

Comment:
`Type` in that case is `java.lang.reflect.Type`, this can be a `Class` or the wrap type `SimpleType` I created to wrap more information in case of map/collections. `Types.of()` will check the argument type and introspect accordingly, if it's an subclass of `SimpleType` it just return it, but if it's a instance of `Class` or `ParameterizedType` it do some other stuff.